src/Type/Php/ArrayCombineFunctionReturnTypeExtension.php

```diff
@@ -0,0 +1,106 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\Variable;
+use PHPStan\Analyser\Scope;
+use PHPStan\Php\PhpVersion;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\Type;
+use PHPStan\Type\UnionType;
+
+class ArrayCombineFunctionReturnTypeExtension implements \PHPStan\Type\DynamicFunctionReturnTypeExtension
+{
+
+	private PhpVersion $phpVersion;
+
+	public function __construct(PhpVersion $phpVersion)
+	{
+		$this->phpVersion = $phpVersion;
+	}
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'array_combine';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if (count($functionCall->args) < 2) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$args0 = $functionCall->args[0]->value;
+		$args1 = $functionCall->args[1]->value;
+
+		$keysParamType = $scope->getType($args0);
+		$valuesParamType = $scope->getType($args1);
+
+		if (
+			$keysParamType instanceof ConstantArrayType
+			&& $valuesParamType instanceof ConstantArrayType
+		) {
+			$keyTypes = $keysParamType->getValueTypes();
+			$valueTypes = $valuesParamType->getValueTypes();
+
+			if (count($keyTypes) !== count($valueTypes)) {
+				return new ConstantBooleanType(false);
+			}
+
+			$keyTypes = $this->sanitizeConstantArrayKeyTypes($keyTypes);
+			if ($keyTypes !== null) {
+				return new ConstantArrayType(
+					$keyTypes,
+					$valueTypes
+				);
+			}
+		}
+
+		$arrayType = new ArrayType(
+			$keysParamType instanceof ArrayType ? $keysParamType->getItemType() : new MixedType(),
+			$valuesParamType instanceof ArrayType ? $valuesParamType->getItemType() : new MixedType()
+		);
+
+		if ($this->phpVersion->getVersionId() >= 80000) {
```

Comment:
Please add a describing method to PhpVersion class and use it here.

Reply:
I created `throwsTypeErrorForInternalFunctions`